meetings/working-groups/collection-literals/CL-2023-06-12.md

```diff
@@ -0,0 +1,161 @@
+{
+    public readonly Type FactoryType;
+    public readonly string MethodName;
+}
+```
+
+When present on a type `T`, this identifies a static factory method `MethodName` found on type `FactoryType` which can be used to construct an instance of `T`. Several factory shapes will be supported.  Note: The type `T` and the `FactoryType` are allowed to be the same type.  These shapes may also be expanded upon in the future if motivating scenarios exist for them.
+
+### Shape 1: `ReadOnlySpan<T>` as input
+
+The first shape we support is the following:
+
+```c#
+[CollectionBuilder(typeof(ImmutableHashSet), "Create")]
+public class ImmutableHashSet<T> { /*...*/ }
+
+public static class ImmutableHashSet
+{
+    public static ImmutableHashSet<T> Create<T>(ReadOnlySpan<T> values);
+}
+```
+
+Here, the factory method takes a `ReadOnlySpan` of values and returns an instance of the final collection.  The compiler would then translate in the following fashion:
+
+```c#
+ImmutableHashSet<int> values = [1, 2, 3];
+
+// translates to:
+ReadOnlySpan<int> __values = [1, 2, 3];
+ImmutableHashSet<int> values = ImmutableHashSet.Create<int>(__values);
+```
+
+This pattern will allow pointing at methods that already exist for this purpose, as well as allowing the runtime to add new methods to some existing types to allow them to be created from spans.  A similar shape that we want to support is practically the same as the above, except that the factory method can be a constructor of the type.  Continuing the example of above, that would be:
+
+```c#
+[CollectionBuilder(typeof(ImmutableHashSet<>), ".ctor")]
+public class ImmutableHashSet<T>
+{
+    public ImmutableHashSet(ReadOnlySpan<T> values);
+}
+
+// code:
+ImmutableHashSet<int> values = [1, 2, 3];
+
+// translates to:
+ReadOnlySpan<int> __values = [1, 2, 3];
+ImmutableHashSet<int> values = new ImmutableHashSet<int>(__values);
+```
+
+#### Shape 1 consequences
+
+The above approach is desired by the runtime so they can add overloads to their existing construction methods that allow the more efficient use of spans over more expensive collection types (arrays, IEnumerables, etc.).  However, this raises a problem given that these types normally have existing construction methods that take an `IEnumerable<T>`.  Consider, for example:
+
+```c#
+public class HashSet<T>
+{
+    // existing constructor
+    public HashSet(IEnumerable<T> values);
+
+    // newly added
+    public HashSet(ReadOnlySpan<T> values);
+}
+
+// code:
+new HashSet<int>(new int[] { 1, 2, 3 });
+```
+
+With the rules of the language today, the above is ambiguous.  This is unfortunate as a prime reason to create Span overloads of methods is to allow for efficient options for callers that do not need to use the heap.  We believe this should be fixed, and the WG will present a proposal to the LDM to add overload resolution rules that consider Spans better than items in the inheritance chain for arrays.
```

Comment:
```suggestion
With the rules of the language today, the above is ambiguous.  This is unfortunate as a prime reason to create Span overloads of methods is to allow for efficient options for callers that do not need to use the heap.  We believe this should be fixed, and the WG will present a proposal to the LDM to add overload resolution rules that prefers `ReadOnlySpan<T>` over `IEnumerable<T>` when both are applicable.
```

Reply:
hrmm... i think we'd always to prefer ReadOnlySpan over things like IList/etc.   Or ReadOnlySpan over 'object'. etc.

Reply:
I still feel shaky about this rule for 2 reasons:
1. It suggests making a user-defined conversion better than a standard implicit reference conversion. It feels like that's something we have avoided to date.
2. I don't see how you ship APIs that depend on this, without just saying: you can't use this from old C#, sorry.